### ISSUE TITLE
Semantic clash with `Base.insert!`

### ISSUE THREAD

**Issue by c42f** [andyferris/Dictionaries.jl#16]
`insert!(dict, k, val)` with a dictionary means something very different from `insert!` with a `Vector`. The Vector API from Base is more or less sequence- or iterable-like: insert something at a given index of the sequence, and move the remaining values to make room, preserving iteration order.

I wondered whether this is an oversight, or whether there's a plan to deal with this in some way?

**Comment by tkf** [comment]
How about `push!(dict, k => v)`?  It works with `AbstractDict` and `AbstractSet`:

```julia
julia> push!(Dict(), :a => 1)
Dict{Any,Any} with 1 entry:
  :a => 1

julia> push!(IdDict(), :a => 1)
IdDict{Any,Any} with 1 entry:
  :a => 1

julia> push!(Set(), :a)
Set{Any} with 1 element:
  :a

julia> push!(Base.IdSet(), :a)
Base.IdSet{Any} with 1 element:
  :a
```

**Comment by c42f** [comment]
`push!` seems approximately right. Does it matter that `AbstractDictionary` is not really a container of pairs though? This may lead to oddities, such as after `push!(d, e)`, `collect(d)` will not contain `e`?

**Comment by tkf** [comment]
> This may lead to oddities, such as after `push!(d, e)`, `collect(d)` will not contain `e`?

Ah, very good point. Actually, `insert!` looks kind of appropriate now (at least within the functions defined in `Base`) for me.

> insert something at a given index of the sequence, and move the remaining values to make room, preserving iteration order.

I think `insert!(::AbstractDictionary, i, v)` does the first point right. I don't think the last point is really necessary as iteration order is (usually) arbitrary for dictionaries.  The second point ("move the remaining values") is unfortunate but maybe it can be compared to the case where you don't need to "make room"?

```julia
julia> insert!([:a, :b, :c], 4, :d)
4-element Array{Symbol,1}:
 :a
 :b
 :c
 :d
```

**Comment by tkf** [comment]
Reading https://github.com/JuliaLang/julia/pull/35513 and https://github.com/JuliaLang/julia/pull/36070, I wonder if vectors should implement `insertat!` and let `AbstractDictionary` and `AbstractDict` define `insert!`. (With the compatibility fallback for `insert!(vector, ...)`, of course.)

**Comment by c42f** [comment]
Good point. The semantic mismatch between `Base` `insert!` and `delete!` seems very unfortunate.

**Comment by andyferris** [comment]
My thinking here is that `Vector` should use `insertat!`

**Comment by c42f** [comment]
Agreed. Inserting-into-a-sequence and implicitly shifting the keys (indices) of the rest of the sequence seems like a fairly specialized operation. Not something that many kinds of collection will want to implement, so having a specific rather than generic name for it would be an improvement.
